torch/_dynamo/variables/user_defined.py

```diff
@@ -845,11 +845,30 @@ def odict_getitem(self, tx, key):
 
         # TODO this should probably be merged with the dict handling
 
-        index = (
-            key.source
-            if is_hashable(key) and key.source is not None
-            else key.as_python_constant()
-        )
+        if key.source and not is_hashable(key):
+            raise ValueError("Trying to use a non-hashable key in a dict")
+
+        # TODO(anijain2305) Is there a better way to insert guard than iterating through hashable variable trackers?
+        if key.source:
```

Comment:
What about tensors?

Reply:
So, I don't know how to handle tensors here. The example I am dealing is this 

~~~
import torch
x = torch.zeros(3)
def nothing():
    pass

d = {
    1: torch.float32,
    nothing: torch.float32, # Force one of the keys to be not hashable
    x: torch.float32,
}

@torch.compile(backend="eager")
def f(x, y1):
    return torch.zeros(5, dtype=d[y1])

# f(torch.zeros(4), 1)
f(torch.zeros(4), x)
# f(torch.zeros(4), nothing)
~~~

Here `nothing` as key prevents the dict to be converted to `ConstDictVariable`. Instead it is represented as the `UserObjectVariable` tracker. Dynamo creates a `ODictGetItemSource` at `dtype=d[y1]` (odict is just a bad name, there is nothing about ordered dict here).  

Now if `y1` is actually a constant, no problem. 

But in the above example, I made it `TensorVariable`. The problem is what to put in the index of `ODictGetItem`. I want the index to be a PyObject so that I can just save that in the Guard Tree. I thought of using `ConstDictKeySource` by finding the order number of the key in the original dict. But I don't have the original tensor in the TensorVariable, just the `example_value`, so no way to find the order number in the original dict.